### ISSUE TITLE
Support for undici tracing

### ISSUE THREAD

**Issue by flovilmart** [DataDog/dd-trace-js#1615]
Hi there!

We're starting to shift away from axios / request and wish to roll out undici in our new deployments.
Unfortunately, there is no 'official' tracing package for it.

Is it something that is being worked on? Would a PR be accepted to add an `undici` plugin? Should I be aware of any 🐉 ?

Thanks!

**Comment by rochdev** [comment]
> Is it something that is being worked on?

We've been looking into using Undici for sending traces, but we haven't looked into writing a plugin for it yet.

> Would a PR be accepted to add an undici plugin? 

Definitely accepted and welcomed!

> Should I be aware of any 🐉 ?

Not that I know of, it should be pretty similar to other HTTP servers/clients that are already supported.

**Comment by flovilmart** [comment]
Thanks @rochdev ; I'll draw some inspiration from the `http` tracing plugin.

**Comment by bengl** [comment]
@flovilmart Note also that undici has also recently added support for diagnostics channel, so we'll want to use that in versions of undici where that's available.

https://github.com/nodejs/undici/blob/main/docs/api/DiagnosticsChannel.md

**Comment by flovilmart** [comment]
That's a super neat feature those diagnostics channels!
I assume we want both implementations and patch based on the `node.js` and `undici` versions?

**Comment by rochdev** [comment]
@flovilmart It should be based on the `undici` version, but the Node version shouldn't matter since in older versions the module will not exist so the plugin will simply never run. Since we're in the process of rewriting a lot of how plugins work and there are no examples currently of using diagnostics channel in a plugin, I would recommend writing the plugin like any other plugin (for example HTTP) and not worry about diagnostics channel for now.

**Comment by flovilmart** [comment]
OK that works  for me @rochdev , we're not even on node 16 yet! I planned to get on  that next week.

**Comment by flovilmart** [comment]
Got this https://github.com/nodejs/undici/pull/1053 patch done for undici for better compatibility.

Outside the `request` call, patching all other API's (pipeline, stream etc...) is proven to be cumbersome.

The implementation through the diag channels is very straightforward with some book-keeping.
I'll visit the implementation through patching the `Client` and `Request` objects after.

**Comment by flovilmart** [comment]
@rochdev , what do you think about this:https://github.com/nodejs/undici/pull/1053#issuecomment-934560506?

**Comment by rochdev** [comment]
@flovilmart Actually it's fine if the implementation in Undici is backed 100% by diagnostics channel. My thought was that we need to do the monkey-patching anyway to support older versions of Undici without the channel available. In that sense, the idea was to first implement the monkey-patching and then add support to use the channel once it lands in Undici. This doesn't mean adding any monkey-patching ability to Undici since that would be redundant when the channels are available.

So basically:

* dd-trace would support older versions of Undici with monkey-patching.
* dd-trace would support newer versions of Undici with diagnostics channel subscribers when the publishers are implemented in Undici.

In both cases, no change is needed in Undici other than adding diagnostics channel support.

**Comment by JustinTRoss** [comment]
The anticipation is killing me! How did it go @flovilmart ?

**Comment by flovilmart** [comment]
@JustinTRoss I haven't had a chance to actually move forward with a new implementation; the monkey patching look very complex given undici's architecture; there are multiple APIs that leverage dispatchers differently, I haven't found any "clean" monkey patching strategy for all APIs (request, client, pipeline etc...)

**Comment by Grmiade** [comment]
Any news on this subject?
Since the last versions of NodeJS  use Undici to implement the built-in Fetch API, it could be great to have this support =)

**Comment by flovilmart** [comment]
@Grmiade yep agreed - I haven't had the chance to improve on the current design and address the comments.

**Comment by rochdev** [comment]
We're actually in the process of finalizing the approach of instrumenting with `async_hooks` and `diagnostics_channel` now that we have rewritten 50+ plugins and put it to the test, so might be worth it waiting a few weeks so that this can be added in Undici directly. We're planning to implement the new API directly in Node core and ship a polyfill.

**Comment by kibertoad** [comment]
@flovilmart Maybe it might make sense to skip the complex monkey-patching altogether and only support newer versions of Undici?

**Comment by rochdev** [comment]
Quick update related to the above: we're currently finishing up the new approach which we'll be trying internally in the next few weeks, and we plan to add instrumentation to Undici directly in Node core soon after. That work can be tracked in https://github.com/nodejs/node/pull/44943

**Comment by flovilmart** [comment]
Amazing work @rochdev  glad your team was able to pull it through! That was not a simple endeavour!

**Comment by mrkosima** [comment]
hello there, are there any updates on Undici support?

**Comment by SimpleCreations** [comment]
I see that support for `fetch` has been added [here (#3258)](https://github.com/DataDog/dd-trace-js/pull/3258).
Does this mean this issue can be closed now?

**Comment by cassidycodes** [comment]
We tested out undici for our gateway application and datadog didn't report any traces for backend services when we used undici. I don't think this issue can be closed until dd-trace supports undici.

We are using dd-trace-js v 4.14.0

**Comment by max-recharge** [comment]
I meant to say that most people want undici support because the Fetch API uses undici. But I guess there are cases when one might use undici without `fetch`.

**Comment by SimpleCreations** [comment]
I meant to say that most people want undici support because the Fetch API uses undici. But I guess there are cases when one might use undici without `fetch`.

**Comment by dpnova** [comment]
https://datadoghq.dev/dd-trace-js/interfaces/export_.plugins.undici.html

Whatever this is, it doesn't work.

**Comment by alyahmedaly** [comment]
@tlhunter do you mind shared an update about the status, and is there is anything we can do like using hooks to get something working for now?

**Comment by rochdev** [comment]
Closing as support for undici and fetch have landed in 5.17.0 and 4.3.0 respectively.

Note: this is only true for the `fetch` API in both cases. If you are looking for undici's other APIs such as `request`, please upvote #5268
